### ISSUE TITLE
[BUG] ZIO HTTP interpreter potential regression and documentation divergence

### ISSUE THREAD

**Issue by calvinlfer** [softwaremill/tapir#2764]
Tapir version: 1.2.9
ZIO HTTP version: 0.0.4
Scala version: 2.13.x/3.2.x

**Describe the bug**
Given a custom path parameter like so:
```scala
import sttp.tapir.{Codec, CodecFormat, DecodeResult, Endpoint}
final case class Example private (value: String)
object Example {
    given exampleCodec: Codec[String, Example, CodecFormat.TextPlain] = Codec.string.mapDecode(raw =>
      Example.make(raw) match
        case Left(error)   => DecodeResult.Error(raw, new IllegalArgumentException(s"Invalid Example value ($raw), failed with $error"))
        case Right(result) => DecodeResult.Value(result)
    )(_.value)

    def make(in: String): Either[String, Example] =
      if (in.length > 5) Right(new Example(in))
      else Left("Too short")
  }
  ```
And the following endpoint
```scala
  val exampleEndpoint: Endpoint[Unit, Example, String, String, Any] = endpoint.get
    .in("test" / path[Example]("testId"))
    .out(stringBody)
    .errorOut(stringBody)

  val logic: ZServerEndpoint[Any, Any] =
    exampleEndpoint.zServerLogic(example => ZIO.succeed(s"Hello $example"))

  val app = ZioHttpInterpreter().toApp(logic)
```   

Calling this endpoint with a path parameter that does not meet the criteria such as `GET /test/1` would result in a HTTP status 404 in Tapir version 1.2.8 with ZIO HTTP would result in a 404. 

Now with Tapir 1.2.9 
```scala
val app = ZioHttpInterpreter().toHttp(logic)
```

`GET /test/1` now returns status 500. 
 
**What is the problem?**
According to the [documentation](https://tapir.softwaremill.com/en/latest/server/errors.html?highlight=error%20handling#default-failure-handler) this should be a status 404 as it were before since the default decoder failure handler would result in a no-match which is a deviation from the documentation. This is understandable given the change that was made to [`ZioHttpInterpreter.scala`](https://github.com/softwaremill/tapir/compare/v1.2.8...v1.2.9) since `fromOptionalHandlerZIO` is no longer being used. 

The fix to work around this by overriding the `DefaultDecodeFailureHandler` does work but there is gotcha. Doing the following will not work which the documentation does (but using another interpreter)
```scala
val errorHandler: DefaultDecodeFailureHandler =
    DefaultDecodeFailureHandler(
      DefaultDecodeFailureHandler.respond(
        _,
        badRequestOnPathErrorIfPathShapeMatches = true,
        badRequestOnPathInvalidIfPathShapeMatches = true
      ),
      FailureMessages.failureMessage,
      DefaultDecodeFailureHandler.failureResponse
    )

val brokenOptions = ZioHttpServerOptions.default[Any].copy(decodeFailureHandler = errorHandler)
val app = ZioHttpInterpreter(brokenOptions).toHttp(logic)
```

This seems to be because `decodeFailureHandler` is not used directly by the interpreter but instead relies on `createOptions` to set it up properly so using `customiseInterceptors` works:
```scala
  val app = ZioHttpInterpreter(
    ZioHttpServerOptions.customiseInterceptors
      .decodeFailureHandler(errorHandler)
      .options
  ).toHttp(logic)
```

**How to reproduce?**
Here is the full code sample for your reference:
```scala
import sttp.tapir.{Codec, CodecFormat, DecodeResult, Endpoint}
import zio.*
import sttp.tapir.json.circe.*
import sttp.tapir.server.interceptor.decodefailure.DefaultDecodeFailureHandler
import sttp.tapir.server.interceptor.decodefailure.DefaultDecodeFailureHandler.FailureMessages
import sttp.tapir.server.ziohttp.{ZioHttpInterpreter, ZioHttpServerOptions}
import sttp.tapir.ztapir.*
import zio.http.{HttpApp, Server}

object TapirReproducer extends ZIOAppDefault {
  final case class Example private (value: String)
  object Example {
    given exampleCodec: Codec[String, Example, CodecFormat.TextPlain] = Codec.string.mapDecode(raw =>
      Example.make(raw) match
        case Left(error)   => DecodeResult.Error(raw, new IllegalArgumentException(s"Invalid Example value ($raw), failed with $error"))
        case Right(result) => DecodeResult.Value(result)
    )(_.value)

    def make(in: String): Either[String, Example] =
      if (in.length > 5) Right(new Example(in))
      else Left("Too short")
  }

  val exampleEndpoint: Endpoint[Unit, Example, String, String, Any] = endpoint.get
    .in("test" / path[Example]("testId"))
    .out(stringBody)
    .errorOut(stringBody)

  val logic: ZServerEndpoint[Any, Any] =
    exampleEndpoint.zServerLogic(example => ZIO.succeed(s"Hello $example"))

  val errorHandler: DefaultDecodeFailureHandler =
    DefaultDecodeFailureHandler(
      DefaultDecodeFailureHandler.respond(
        _,
        badRequestOnPathErrorIfPathShapeMatches = true,
        badRequestOnPathInvalidIfPathShapeMatches = true
      ),
      FailureMessages.failureMessage,
      DefaultDecodeFailureHandler.failureResponse
    )

  val brokenOptions  = ZioHttpServerOptions.default[Any].copy(decodeFailureHandler = errorHandler)
  val workingOptions = ZioHttpServerOptions.customiseInterceptors[Any].decodeFailureHandler(errorHandler).options

  val app = ZioHttpInterpreter(workingOptions).toHttp(logic)

  val run: ZIO[ZIOAppArgs, Any, Any] =
    Server.serve(app.withDefaultErrorResponse).provide(Server.default)
}
```
Replace `brokenOptions` with `workingOptions` to see the status 500 behavior.

**Additional information**
To summarize there are 2 issues:
* the documentation seems to have diverged (documentation says 404 by default but now we are getting 500) - not sure if this okay. Tapir 1.2.8 correctly produced 404 but 1.2.9 produces 500
* using copy to build out the `DefaultDecodeFailureHandler` on `ZioHttpServerOptions` does not result in the intended configuration

**Comment by kciesielski** [comment]
@calvinlfer the handler indeed shouldn't be a member of `ZioHttpServerOptions`, I have just removed it from there.
